### PR TITLE
HDDS-10517. Recon - Add a UI component for showing DN decommissioning detailed status and info

### PR DESCRIPTION
**Pull request by smitajoshi12**
## What changes were proposed in this pull request?
UI component has been added  with some drill down on a DN which is under decommissioning. 

Backend PR For Reference:-
https://github.com/apache/ozone/pull/6376## 

What is the link to the Apache JIRA
(https://issues.apache.org/jira/browse/HDDS-10517)

## How was this patch tested?
Manually

New Enhancemnet Tested with Cluster data in local
1)  Added New card  with DataNodes Decommission Count

image

3) After Hovering on UUID Column
image

4) Tooltip has been added before hover
image

5) Once Status becomes Decommissioned  decommission details will not be shown.
image

6)  From Backend if we get Empty response from decommission api will not showing any details about decommission.

7) Handled all negative test cases for empty and null data if one Api's fail other apis working

image

8) With Multiple  datanode Decommission

image

image

9) After Error on Card 

image

10) Vertical Layout 

image

### CODE REVIEW

**Comment by devmadhuu** [comment]
@smitajoshi12 thanks for working on this patch. Looks like on datanodes page, you wait for /`datanodes` API to provide the info if any datanode is in decommissioning or not, then you call the decommission info API for that respective UUID, which seems inconsistent behavior, because on overview summary page, you are showing the count of datanodes which are in decommissioning , but when user clicks on summary tile, and taken to datanodes page, you still show all nodes healthy which is wrong. Kindly correct the behavior and make it consistent.

**Comment by smitajoshi12** [comment]
> @smitajoshi12 thanks for working on this patch. Looks like on datanodes page, you wait for /`datanodes` API to provide the info if any datanode is in decommissioning or not, then you call the decommission info API for that respective UUID, which seems inconsistent behavior, because on overview summary page, you are showing the count of datanodes which are in decommissioning , but when user clicks on summary tile, and taken to datanodes page, you still show all nodes healthy which is wrong. Kindly correct the behavior and make it consistent.

@devmadhuu  It is taking time to reflect status in /datanodes page. /datanode api is called after  each 60 seconds as default behaviour. will think how we can meet this condition.

**Comment by smitajoshi12** [comment]
@devmadhuu 
Addressed synchronization isssue in latest commit and attached new screenshots.

File: hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/api/db.json [hunk 1]
```diff
@@ -742,55 +742,55 @@
...
-              "name": "STANDALONE",
-              "value": 9859
-            }
+            "level": 0,
+            "parent": null,
+            "cost": 0,
+            "uuid": "9e4e92da-4abd-461d-b042-dc4fe72e8727",
+            "uuidString": "9e4e92da-4abd-461d-b042-dc4fe72e8727",
+            "ipAddress": "172.22.0.5",
+            "hostName": "ozone_datanode_1.ozone_default",
```
Comment (ArafatKhan2198) [inline hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/api/db.json:751]: Could you please revert these changes from the code.
The changes show only tab spaces being added.

File: hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/api/db.json [hunk 2]
```diff
@@ -22,7 +22,7 @@
     "datanodes": [
       {
         "hostname": "localhost2.storage.enterprise.com",
-        "uuid": "b590734e-a5f2-11ea-bb37-0242ac130002",
+        "uuid": "4712ba3d-4bb2-477a-9211-d9b50c013055",
```
Comment (ArafatKhan2198) [inline hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/api/db.json:25]: Why did we change the UUID's here?

File: hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/datanodes/decommissionSummary.tsx [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+ */
+
+import React from 'react';
+import { Descriptions, Popover, Tooltip, Icon } from 'antd';
+import { withRouter } from 'react-router-dom';
+import { RouteComponentProps } from 'react-router';
+import axios from 'axios';
+import { showDataFetchError } from 'utils/common';
+
+interface IDecommissionSummaryProps extends RouteComponentProps<object> {
```
Comment (ArafatKhan2198) [inline hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/datanodes/decommissionSummary.tsx:26]: Could you please provide a short usage summary of why this is added.

**Review by ArafatKhan2198 — COMMENTED** [review]
Thanks for working on this smitha!
Have posted a few comments and I also have a few doubts on the additions you have made can you please take a look.
And also please resolve the conflict

File: hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/api/db.json (hunk 2 above)
Comment (smitajoshi12) [reply hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/api/db.json:25]: @ArafatKhan2198  This for Mock Data in Local to show Tooltip as it is going to decommission soon or local testing.

**Comment by devmadhuu** [comment]
@smitajoshi12 we shouldn't show the initial tool tip from datanodes page "`It is going to decommission soon!!!`". Rather we should call decommission info API and match with what all datanodes are in decommissioning and update their status on UI. Behavior should be consistent across all UI pages for same entity. And also you need to call the decommission info API once , not multiple times as user land up on datanodes page.

**Comment by smitajoshi12** [comment]
> @smitajoshi12 we shouldn't show the initial tool tip from datanodes page "`It is going to decommission soon!!!`". Rather we should call decommission info API and match with what all datanodes are in decommissioning and update their status on UI. Behavior should be consistent across all UI pages for same entity. And also you need to call the decommission info API once , not multiple times as user land up on datanodes page.

@devmadhuu 
Issue with Calling decommision API once is datanode api is called after 1 min if we call decommission API only once then we will not get updated values so checking UUIDs from decommission API and comparision will give incorrect results as i checked with Cluster data. 
It is going to decommmission soon msg is updated after calling decommision API and datanode API as we discussed previously datanode api is taking time to update records.

**Comment by devmadhuu** [comment]
> > @smitajoshi12 we shouldn't show the initial tool tip from datanodes page "`It is going to decommission soon!!!`". Rather we should call decommission info API and match with what all datanodes are in decommissioning and update their status on UI. Behavior should be consistent across all UI pages for same entity. And also you need to call the decommission info API once , not multiple times as user land up on datanodes page.
> 
> @devmadhuu Issue with Calling decommision API once is datanode api is called after 1 min if we call decommission API only once then we will not get updated values so checking UUIDs from decommission API and comparision will give incorrect results as i checked with Cluster data. It is going to decommmission soon msg is updated after calling decommision API and datanode API as we discussed previously datanode api is taking time to update records.

@smitajoshi12 As discussed, pls change the operational state column value based on decommission info API on datanodes page. Showing tooltip is wrong because datanode decommissioning has already started.

**Comment by smitajoshi12** [comment]
> > > @smitajoshi12 we shouldn't show the initial tool tip from datanodes page "`It is going to decommission soon!!!`". Rather we should call decommission info API and match with what all datanodes are in decommissioning and update their status on UI. Behavior should be consistent across all UI pages for same entity. And also you need to call the decommission info API once , not multiple times as user land up on datanodes page.
> > 
> > 
> > @devmadhuu Issue with Calling decommision API once is datanode api is called after 1 min if we call decommission API only once then we will not get updated values so checking UUIDs from decommission API and comparision will give incorrect results as i checked with Cluster data. It is going to decommmission soon msg is updated after calling decommision API and datanode API as we discussed previously datanode api is taking time to update records.
> 
> @smitajoshi12 As discussed, pls change the operational state column value based on decommission info API on datanodes page. Showing tooltip is wrong because datanode decommissioning has already started.

@devmadhuu  
I have done changes suggested by you update column manually as comparing uuids from decommission api. If one api fails it will not block other apis attached screenshots.
2 Apis handled synchronously.

**Comment by devmadhuu** [comment]
@smitajoshi12 kindly resolve the conflicts in your code.

**Comment by smitajoshi12** [comment]
> @smitajoshi12 kindly resolve the conflicts in your code.



> @smitajoshi12 kindly resolve the conflicts in your code.

@devmadhuu 
Resolved conflicts and added new screenshot of Overview Page  as alignment for tile has been changed.

**Comment by smitajoshi12** [comment]
> @smitajoshi12 kindly resolve the conflicts in your code.

@devmadhuu 
Hi Devesh Resolved conflicts and merged master with all changes and did testing and pushed changes in latest commit.

**Comment by smitajoshi12** [comment]
@dombizita @devabhishekpal 
Hi Zita and Abhishek Could you review this PR

**Comment by smitajoshi12** [comment]
> @smitajoshi12 Pls resolve your conflicts.

Resolved Merge Conflicts

**Comment by ArafatKhan2198** [comment]
Can you finish all the pending changes on this patch @smitajoshi12 
Make sure you go through all of them

**Comment by smitajoshi12** [comment]
> Can you finish all the pending changes on this patch @smitajoshi12 Make sure you go through all of them

@ArafatKhan2198 
Completed all changes and sugggestion suggested and pushed code.

File: hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/api/db.json (hunk 1 above)
Comment (ArafatKhan2198) [reply hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/api/db.json:753]: These can be reverted right @smitajoshi12

File: hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/datanodes/datanodes.tsx [hunk 4]
```diff
@@ -190,15 +190,21 @@ const COLUMNS = [
...
     key: 'uuid',
     isVisible: true,
     isSearchable: true,
     sorter: (a: IDatanode, b: IDatanode) => a.uuid.localeCompare(b.uuid),
-    defaultSortOrder: 'ascend' as const
+    defaultSortOrder: 'ascend' as const,
+    render: (uuid: IDatanode, record: IDatanode) => {
+      return (
+        //1. Compare Decommission Api's UUID with all UUID in table and show Decommission Summary
+        (decommissionUids && decommissionUids.includes(record.uuid) && record.opState !== 'DECOMMISSIONED') ?
```
Comment (devabhishekpal) [inline hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/datanodes/datanodes.tsx:204]: Minor nit for UUID spelling
```suggestion
        (decommissionUuids && decommissionUuids.includes(record.uuid) && record.opState !== 'DECOMMISSIONED') ? 
```

File: hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/datanodes/datanodes.tsx [hunk 5]
```diff
@@ -389,24 +398,39 @@ export class Datanodes extends React.Component<Record<string, object>, IDatanode
...
 
-  _loadData = () => {
-    this.setState(prevState => ({
-      loading: true,
-      selectedColumns: this._getSelectedColumns(prevState.selectedColumns)
-    }));
+  _loadData = async () => {
+    // Need to call decommission API on each interval to get updated status before datanode API to compare UUID's
+    // update Operation State Column in table Manually before rendering
+    let decommissionAPIRequest = this._loadDecommisionAPI();
```
Comment (devabhishekpal) [inline hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/datanodes/datanodes.tsx:404]: I think we don't need both `await` and `then()`

```suggestion
    let decommissionAPIResponse = await this._loadDecommisionAPI();
    decomissionUids = decommissionAPIResponse && decommissionAPIResponse.data
```

File: hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/overview/overview.tsx [hunk 6]
```diff
@@ -304,6 +309,8 @@ export class Overview extends React.Component<Record<string, object>, IOverviewS
           <Col xs={24} sm={18} md={12} lg={12} xl={6}>
             <OverviewCard loading={loading} title='Deleted Containers' data={deletedContainers.toString()} icon='delete' />
           </Col>
+        </Row>
+        <Row gutter={[25, 25]}>
```
Comment (devabhishekpal) [inline hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/overview/overview.tsx:313]: We don't need to separate out into rows.
Single row parent wrapper would automatically handle card breaks to new rows.
Could you check once without this <Row> component wrapped?

**Review by devabhishekpal — CHANGES_REQUESTED** [review]

File: hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/overview/overview.tsx [hunk 7]
```diff
@@ -320,6 +327,10 @@ export class Overview extends React.Component<Record<string, object>, IOverviewS
               <OverviewCard title="OM Service" loading={loading} data={omServiceId} icon='file-text' linkToUrl='/Om' />
             </Col>
           }
+          <Col xs={24} sm={18} md={12} lg={12} xl={6}>
+            <OverviewCard loading={loading} title='Decommissioning Datanodes Summary' data={decommissionInfoCount.toString()} icon='hourglass'
```
Comment (devabhishekpal) [inline hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/overview/overview.tsx:331]: Since icons are not supported like this in AntD v4, we need to specify the icon in OverviewCard component as well.
Check the following function and add the hourglass optionn in that:
https://github.com/apache/ozone/blob/975a8d85a343a3b7a971b6798493280684456136/hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/components/overviewCard/overviewCard.tsx#L64

**Review by devabhishekpal — CHANGES_REQUESTED** [review]

File: hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/datanodes/datanodes.tsx (hunk 4 above)
Comment (smitajoshi12) [reply hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/datanodes/datanodes.tsx]: Done changes in latest 2 commits

File: hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/datanodes/datanodes.tsx (hunk 5 above)
Comment (smitajoshi12) [reply hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/datanodes/datanodes.tsx]: Done changes in latest 2 commits

File: hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/overview/overview.tsx (hunk 6 above)
Comment (smitajoshi12) [reply hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/overview/overview.tsx]: Done changes in latest 2 commits.

File: hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/overview/overview.tsx (hunk 7 above)
Comment (smitajoshi12) [reply hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/overview/overview.tsx:329]: @devabhishekpal 
I merged code with latest master I could not see const IconSelector = ({ iconType, ...extras }: { iconType: string }) => { 

Can we do in seprate pr or we can have discussion when to add changes

**Comment by devmadhuu** [comment]
@smitajoshi12 Thanks for improving the patch. Changes LGTM +1

**Comment by devmadhuu** [comment]
@devabhishekpal  we need that minimal information for a node which is in decommission stage.

**Comment by devabhishekpal** [comment]
Thanks for letting me know @devmadhuu.
Then I guess we can go with a more vertical layout as I attached in the design file @smitajoshi12.

**Comment by smitajoshi12** [comment]
> @devabhishekpal we need that minimal information for a node which is in decommission stage.

@devabhishekpal 
https://issues.apache.org/jira/browse/HDDS-10514) I have refered this JIRA for details. Will Check Vertical Layout as it is collapsable if we have less information.

**Comment by smitajoshi12** [comment]
> Hi @smitajoshi12 , so I was suggesting to change the component layout as below: Frame 1
> 
> This would make sure we are not covering too much information for the DN with the popup Also could you let me know what `No. of Under-Replicated Containers`, `No. of Unclosed Pipelines`, `No. of Unclosed Containers` are used for? If it is not very relevant to the user to get this information, we can omit this info I guess. @devmadhuu @ArafatKhan2198 what do you think?

@devabhishekpal 
Done the changes in latest commit.8b82426c1568815fa88079d2ddfec738f7b85dbd

File: hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/datanodes/decommissionSummary.tsx [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+      isLoading: false,
+      summaryData: []
+    };
+  }
+
+  componentDidMount(): void {
+    this.setState({
+      isLoading: true
+    });
+    if (this.state.record && this.state.summaryData !== 'null' && this.state.summaryData !== 'undefined') {
```
Comment (devabhishekpal) [inline hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/datanodes/decommissionSummary.tsx:49]: I think this can be simplified to:
```suggestion
    if (this.state?.record && this.state?.summaryData) {
```

File: hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/datanodes/decommissionSummary.tsx [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+          {metrics !== null && metrics !== undefined && Object.keys(metrics).length !== 0 &&
+            <>
+              {<Descriptions.Item label="Decommissioning Started at">{metrics.decommissionStartTime}</Descriptions.Item>}
+              {<Descriptions.Item label="No. of Unclosed Pipelines">{metrics.numOfUnclosedPipelines}</Descriptions.Item>}
+              {<Descriptions.Item label="No. of Unclosed Containers">{metrics.numOfUnclosedContainers}</Descriptions.Item>}
+              {<Descriptions.Item label="No. of Under-Replicated Containers">{metrics.numOfUnderReplicatedContainers}</Descriptions.Item>}
+            </>
+          }
+          {
+            containers !== null && containers !== undefined && Object.keys(containers).length !== 0 &&
```
Comment (devabhishekpal) [inline hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/datanodes/decommissionSummary.tsx:93]: Since containers is an object I think this can also be simplified to:
```suggestion
            containers && Object.keys(containers).length !== 0 &&
```

File: hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/datanodes/decommissionSummary.tsx [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+              {containers.UnClosed && containers.UnClosed.length > 0 && <Descriptions.Item label="Unclosed">{containers.UnClosed}</Descriptions.Item>}
+            </>
+          }
+        </Descriptions>
+      );
+    }
+    // Need to check summarydata is not empty 
+    return (
+      <>
+        { (summaryData !== 'null' && summaryData !== 'undefined' && summaryData && summaryData.length !== 0) ?
```
Comment (devabhishekpal) [inline hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/datanodes/decommissionSummary.tsx:105]:
```suggestion
        { (summaryData && summaryData.length !== 0) ?
```

File: hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/overview/overview.tsx [hunk 11]
```diff
@@ -280,6 +288,13 @@ export class Overview extends React.Component<Record<string, object>, IOverviewS
       openKeysError = true;
     }
 
+    if ([
+      decommissionInfoCount].some(
+        (data) => data === undefined
+      )) {
```
Comment (devabhishekpal) [inline hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/overview/overview.tsx:294]: Nit:
```suggestion
    if ([decommissionInfoCount].some(
        (data) => data === undefined
      )) {
```

**Review by devabhishekpal — CHANGES_REQUESTED** [review]
Thanks for the patch @smitajoshi12 , the changes look good to me.
Just a few minor comments for better code quality

**Review by ArafatKhan2198 — APPROVED** [review]
@smitajoshi12, please review the final comments from @devabhishekpal. I have tested the patch, and both the layout and functionality are working well. The patch looks good to me.

File: hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/datanodes/decommissionSummary.tsx (hunk 8 above)
Comment (smitajoshi12) [reply hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/datanodes/decommissionSummary.tsx]: Done the changes in latest commit3e08fb7b48adea2a78e0f77408427da00db453bf and done testing with local cluster data.

File: hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/datanodes/decommissionSummary.tsx (hunk 9 above)
Comment (smitajoshi12) [reply hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/datanodes/decommissionSummary.tsx]: @devabhishekpal 
Done the changes in latest commit https://github.com/apache/ozone/commit/3e08fb7b48adea2a78e0f77408427da00db453bf and done testing with local cluster data.

File: hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/datanodes/decommissionSummary.tsx (hunk 10 above)
Comment (smitajoshi12) [reply hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/datanodes/decommissionSummary.tsx]: Done the changes in latest commit https://github.com/apache/ozone/commit/3e08fb7b48adea2a78e0f77408427da00db453bf and done testing with local cluster data.

File: hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/overview/overview.tsx (hunk 11 above)
Comment (smitajoshi12) [reply hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/overview/overview.tsx]: Done the changes in latest commit https://github.com/apache/ozone/commit/3e08fb7b48adea2a78e0f77408427da00db453bf and done testing with local cluster data.

**Comment by devmadhuu** [comment]
Thanks @smitajoshi12 for improving the patch. Now it LGTM +1

**Review by devabhishekpal — APPROVED** [review]
Thanks for the changes @smitajoshi12 
These changes looks good to me, +1

**Comment by ArafatKhan2198** [comment]
Thanks for the patch @smitajoshi12 & thanks @devabhishekpal @devmadhuu for the review.
